internal/db/id/field.go

```diff
@@ -12,69 +12,160 @@ package id
+	uniqueKey := strconv.Itoa(int(collectionShortID)) + ":" + fieldID
+
+	cache := getFieldShortIDCache(ctx)
+	shortID, ok := cache[uniqueKey]
+	if ok {
+		return shortID, nil
+	}
+
+	// If we miss the cache, load the entire collection's worth - it is almost always
+	// going to be more efficient than loading the field short-ids one by one, and we'll
+	// usually want most of them.
+
+	key := keys.NewFieldIDP(collectionShortID)
+	iter, err := txnctx.MustGet(ctx).Systemstore().Iterator(ctx, corekv.IterOptions{Prefix: key.Bytes()})
+	if err != nil {
+		return 0, err
 	}
 
-	for i := range definitions {
-		fieldSeq, err := sequence.Get(ctx, txn, keys.NewFieldIDSequenceKey(definitions[i].Version.CollectionID))
+	for {
+		hasNext, err := iter.Next()
 		if err != nil {
-			return err
+			return 0, errors.Join(err, iter.Close())
+		}
+		if !hasNext {
+			break
 		}
 
-		for j := range definitions[i].Version.Fields {
-			var fieldID client.FieldID
-			if definitions[i].Version.Fields[j].ID != client.FieldID(0) {
-				fieldID = definitions[i].Version.Fields[j].ID
-			} else if definitions[i].Version.Fields[j].Name == request.DocIDFieldName {
-				// There is no hard technical requirement for this, we just think it looks nicer
-				// if the doc id is at the zero index.  It makes it look a little nicer in commit
-				// queries too.
-				fieldID = 0
-			} else {
-				nextID, err := fieldSeq.Next(ctx, txn)
-				if err != nil {
-					return err
-				}
-				fieldID = client.FieldID(nextID)
-			}
-
-			if definitions[i].Version.Fields[j].Kind.HasValue() {
-				switch kind := definitions[i].Version.Fields[j].Kind.Value().(type) {
-				case *client.NamedKind:
-					var newKind client.FieldKind
-					if kind.Name == definitions[i].Version.Name {
-						newKind = client.NewSelfKind("", kind.IsArray())
-					} else if otherSchema, ok := schemasByName[kind.Name]; ok {
-						newKind = client.NewSchemaKind(otherSchema.Root, kind.IsArray())
-					} else {
-						// Continue, and let the validation stage return user friendly errors
-						// if appropriate
-						continue
-					}
-
-					definitions[i].Version.Fields[j].Kind = immutable.Some(newKind)
-				default:
-					// no-op
-				}
-			}
-
-			definitions[i].Version.Fields[j].ID = fieldID
+		key, err := keys.NewFieldIDFromString(string(iter.Key()))
+		if err != nil {
+			return 0, errors.Join(err, iter.Close())
+		}
+
+		value, err := iter.Value()
+		if err != nil {
+			return 0, errors.Join(err, iter.Close())
+		}
+
+		v, err := strconv.ParseUint(string(value), 10, 0)
+		if err != nil {
+			return 0, err
+		}
+		sID := uint32(v)
+
+		// This concatenation is temporary, soon we can just use the field CID
+		uniqueKey := strconv.Itoa(int(collectionShortID)) + ":" + key.FieldID
+		cache[uniqueKey] = sID
+
+		if key.FieldID == fieldID {
+			shortID = sID
+		}
+	}
+
+	return shortID, iter.Close()
+}
+
+// SetShortFieldID sets and stores the short field id, if it does not already exist.
+func SetShortFieldID(
```

Comment:
suggestion: Unless you know that this will be used outside this package eventually, I suggest making it private.

Reply:
I think it will be, and the func below that consumes it will either be removed or (less likely) moved.